### DISCUSSION TITLE
MQTT error - last_result.log does not exist

### DISCUSSION THREAD

**Discussion by PyrokineticDarkElf** [netalertx/NetAlertX#572] [Q&A]
Not sure if anyone can suggest what I've done wrong here? I've been trying to get this working with Mosquitto (container not in Home Assistant). Whenever I try the test button I seem to get the following error:

```
Traceback (most recent call last):
  File "/home/pi/pialert/front/plugins/_publisher_mqtt/mqtt.py", line 363, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/pi/pialert/front/plugins/_publisher_mqtt/mqtt.py", line 61, in main
    mqtt_start(db)
  File "/home/pi/pialert/front/plugins/_publisher_mqtt/mqtt.py", line 255, in mqtt_start
    client = mqtt_create_client()     
             ^^^^^^^^^^^^^^^^^^^^
  File "/home/pi/pialert/front/plugins/_publisher_mqtt/mqtt.py", line 239, in mqtt_create_client
    client = mqtt_client.Client('PiAlert')   # Set Connecting Client ID    
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/myenv/lib/python3.11/site-packages/paho/mqtt/client.py", line 766, in __init__
    raise ValueError(
ValueError: Unsupported callback API version: version 2.0 added a callback_api_version, see migrations.md for details

22:41:24 [Plugins] ⚠ ERROR - enable LOG_LEVEL=debug and check logs
22:41:24 [Plugins] The file /home/pi/pialert/front/plugins/_publisher_mqtt/last_result.log does not exist
22:41:24 [Plugins] No output received from the plugin MQTT - enable LOG_LEVEL=debug and check logs
22:41:24 [2024-02-17 22:41:24+00:00] END Run: MQTT
22:41:24 [Test] END Test: MQTT
```

Can I confirm the 'topic' I should be using in Home Assistant is 'homeassistant'?:
![image](https://github.com/jokob-sk/Pi.Alert/assets/93349494/8307e5ef-c0b1-4d83-8afa-bd82662bd67c)

Am I meant to be using a specific username for MQTT? "PiAlert":
`client = mqtt_client.Client('PiAlert')   # Set Connecting Client ID`

**Comment by jokob-sk** [comment]
Hi @Staples1010 ,

I haven't tested a stand-alone Mosquitto instance. 

Can you please confirm you are using the correct details for Broker, Port, Username, Password?

https://github.com/jokob-sk/Pi.Alert/blob/main/docs/HOME_ASSISTANT.md

The below are the most important settings:

![image](https://github.com/jokob-sk/Pi.Alert/assets/96159884/2d08a6d5-19c6-4d74-89b5-73a432c82f26)

Please make sure they align with your Mosquito MQTT settings:

https://github.com/jokob-sk/Pi.Alert/tree/main/front/plugins/_publisher_mqtt

![image](https://github.com/jokob-sk/Pi.Alert/assets/96159884/3d9dcbce-f40c-4c3e-8148-1ee5815139c3)


YOu can check the message, including topic, that is sent here:

https://github.com/jokob-sk/Pi.Alert/blob/34f56585161cffdb9a3a3ded76e9da6195e91af8/front/plugins/_publisher_mqtt/mqtt.py#L191

**Reply by jokob-sk** [reply]
Digging further into this it seems like there has been a breaking change in the paho MQTT library: https://eclipse.dev/paho/files/paho.mqtt.python/html/migrations.html

Still investigating

**Reply by PyrokineticDarkElf** [reply]
As far as I can see, my details are correct.

The only things I can think of that are different are the username containing '-' and my Mosquitto having multiple users 'haos' & 'pi-alert'.
I tried using the 'haos' user to see if it made a difference but it doesn't seem to :/

I've also tried my other user for a known working service, copied the IP and port from that service to be sure and it's still giving the same error.

The [Mosquitto container](https://hub.docker.com/_/eclipse-mosquitto) is working for the other service (with Home Assistant).

I've tried creating the missing file `/home/pi/pialert/front/plugins/_publisher_mqtt/last_result.log` but that didn't seem to work... Probably because it had no content, but it was worth a shot 😅

Any suggestions on what I could try? Other than installing Mosquitto on HA...

**Reply by jokob-sk** [reply]
Hey @Staples1010 ,

This was probably related to this: https://github.com/eclipse/paho.mqtt.python/issues/814

I think I fixed it in this commit: https://github.com/jokob-sk/Pi.Alert/commit/d4b701653e8241284fcec0ed6ff8be8e3bd98db5

This should be available in the next release. If you can, please have a look at the `pi.alert_dev` docker image, in about 15 minutes (or after the last [action finishes](https://github.com/jokob-sk/Pi.Alert/actions)) from now. 

It would be great if you could test this on your end by switching to the above image and letting me know if the issue was resolved.

Thanks in advance,
j

**Reply by PyrokineticDarkElf** [reply]
It looks like more is happening when running the test, but I'm still not seeing anything appear in MQTT Explorer :/

The execution queue seems to never clear, they just keep getting added:
![image](https://github.com/jokob-sk/Pi.Alert/assets/93349494/fdbb2186-29c5-43a1-aad8-81c63c08ee87)

It seems there's a new error related to a missing file `/home/pi/pialert/front/api/table_plugins_objects.json`:
*data obfuscated
```
11:37:50 [Plugin utils] ⚠ ERROR - JSONDecodeError or FileNotFoundError for file /home/pi/pialert/front/api/table_plugins_objects.json
11:37:50 [MQTT] New sensor entry mac          : ba:01:c9:80:8c:78
11:37:50 [MQTT] New sensor entry input_string : mac_ba_01_c9_80_8c_78Device Name Heresensoris_newbell-alert-outline
11:37:50 [MQTT] New sensor entry hash_value   : dcee5d9df23d1131606c1a6193d90852
11:37:50 [MQTT] Publishing sensor number 64
11:37:50 [MQTT] Sending MQTT topic: homeassistant/sensor/mac_ba_01_c9_80_8c_78/is_new/config
11:37:50 [MQTT] Sending MQTT message: {                 "name":"is_new",                 "state_topic":"system-sensors/sensor/mac_ba_01_c9_80_8c_78/state",                 "value_template":"{{value_json.is_new}}",                 "unique_id":"mac_ba_01_c9_80_8c_78_sensor_is_new",                 "device":                     {                         "identifiers": ["mac_ba_01_c9_80_8c_78_sensor"],                         "manufacturer": "PiAlert",                         "name":"Device Name Here"                     },                 "icon":"mdi:bell-alert-outline"                 }
11:37:52 [Plugin utils] ⚠ ERROR - JSONDecodeError or FileNotFoundError for file /home/pi/pialert/front/api/table_plugins_objects.json
11:37:52 [MQTT] New sensor entry mac          : ba:01:c9:80:8c:78
11:37:52 [MQTT] New sensor entry input_string : mac_ba_01_c9_80_8c_78Device Name Heresensorvendorcog
11:37:52 [MQTT] New sensor entry hash_value   : b85ad03bc952800de297b23ab4f36e1c
11:37:52 [MQTT] Publishing sensor number 65
11:37:52 [MQTT] Sending MQTT topic: homeassistant/sensor/mac_ba_01_c9_80_8c_78/vendor/config
11:37:52 [MQTT] Sending MQTT message: {                 "name":"vendor",                 "state_topic":"system-sensors/sensor/mac_ba_01_c9_80_8c_78/state",                 "value_template":"{{value_json.vendor}}",                 "unique_id":"mac_ba_01_c9_80_8c_78_sensor_vendor",                 "device":                     {                         "identifiers": ["mac_ba_01_c9_80_8c_78_sensor"],                         "manufacturer": "PiAlert",                         "name":"Device Name Here"                     },                 "icon":"mdi:cog"                 }
```

I've also noticed that when you attempt to update the settings it seems to hang on 'Importing settings and re-initializing...':
![image](https://github.com/jokob-sk/Pi.Alert/assets/93349494/e52640ad-73cb-428a-9ca1-e7bb4ee766c5)

**Reply by PyrokineticDarkElf** [reply]
And somehow I missed this... The devices table is gone 👀 
![image](https://github.com/jokob-sk/Pi.Alert/assets/93349494/6632905e-37b8-4e5c-86ef-a991fd536d63)

**Reply by jokob-sk** [reply]
Hi @Staples1010 ,

Can you have a look if there are errors in the JSON generated by the app?

https://github.com/jokob-sk/Pi.Alert/blob/main/docs/DEBUG_INVALID_JSON.md

Please access the following and run the JSON files thru a JSON validator to check if something isn't causing the data to break.

- `http://<server>/api/table_settings.json`
- `http://<server>/api/table_devices.json`

Please also try opening the app in a new browser or clearing the cache in case that's the issue.

Also, please make sure you wait long enough for the execution queue to finish. One MQTT sync can take up to 10-15 minutes depending on the number of devices (~3-10s per device).

Thanks in advance for the additional info,
j

**Reply by PyrokineticDarkElf** [reply]
Hi @jokob-sk 👋

Sorry for the radio silence, I've just had a chance to pick this up again and It looks like the MQTT notifications are working for me now!

I can see you've been making some updates (and a re-brand) so presumably, something's changed that has helped :D

Thanks for your help!

**Reply by jokob-sk** [reply]
Hey @Staples1010 ,

Thanks for checking in! Glad to hear the issue is resolved :) 

Yes, changed the name just yesterday - hope it helps differentiate between the stale original project and other active forks and people stop asking me if it's only for the Raspberry Pi 😅

What do you think - would it be ok to also rename the MQTT sensor names?

**Comment by PyrokineticDarkElf** [comment]
@jokob-sk It would be okay with me, but If people already have automations in Home Assistant set up using the existing name, could this break their setups?

Having a toggle in the settings could be good, or maybe a textbox for a custom name?

**Reply by jokob-sk** [reply]
Sure, added these two settings for the next release:

![image](https://github.com/jokob-sk/NetAlertX/assets/96159884/282f729d-3f53-4965-aba8-00b8bffa6423)

**Comment by ingoratsdorf** [comment]
The above callback error was fixed in pull request https://github.com/jokob-sk/NetAlertX/pull/759.

The paho library is now v2.0 and will be dropping support for V1 callback.
If a user selected v1 as callback type, the callback in the plugin would fail due to the expected signature being wrong.
My pull requests solved this be eliminating v1 callbacks and offering to rather select the MQTT protocol 3 or 5.
Most modern MQTT brokers will support V5, have tested with EMQX and Mosquitto.
